File: QytetetJava/src/interfaz_textual_qytetet/VistaTextualQytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package interfaz_textual_qytetet;

import java.util.ArrayList;
import java.util.Map;
import java.util.Scanner;
import java.util.TreeMap;
import modeloqytetet.MetodoSalirCarcel;


public class VistaTextualQytetet {
    
private static final Scanner in = new Scanner (System.in);

public void pausa() {
    mostrar("\nPulse Intro");
    in.nextLine();
}
    
public int menuGestionInmobiliaria(){ //ejemplo de menú
     
    this.mostrar("\nElige la gestion inmobiliaria que deseas hacer");
    Map<Integer, String> menuGI = new TreeMap();
    menuGI.put(0, "Siguiente Jugador"); 
    menuGI.put(1, "Edificar casa");
    menuGI.put(2, "Edificar Hotel"); 
    menuGI.put(3, "Vender propiedad ");  	
    menuGI.put(4, "Hipotecar Propiedad"); 
    menuGI.put(5, "Cancelar Hipoteca");
    int salida=this.seleccionMenu(menuGI); // Método para controlar la elección correcta en el menú 
    return salida;
 }

public MetodoSalirCarcel menuSalirCarcel(){    
    this.mostrar("Elige el método para salir de la carcel");
    Map<Integer, String> menuSC = new TreeMap();
    menuSC.put(0, "Salir pagando");
    menuSC.put(1, "Tirar dado");
    int salida=this.seleccionMenu(menuSC);
    

    if (salida == 0)
        return MetodoSalirCarcel.PAGANDOLIBERTAD;
    else
        return MetodoSalirCarcel.TIRANDODADO;
}
  
  public boolean elegirQuieroComprar(){
      this.mostrar("Elegir comprar título de propiedad");
      Map<Integer, String> menuQC = new TreeMap();
      menuQC.put(0, "Quiero comprar");
      menuQC.put(1, "No quiero comprar");
      int salida=this.seleccionMenu(menuQC);
      
      return salida == 0;                   
 }
 
        
 public int menuElegirPropiedad(ArrayList<String> listaPropiedades){  //numero y nombre de propiedades            
    Map<Integer, String> menuEP = new TreeMap();
    menuEP.put(0, "\tNo elegir propiedad");
    int numeroOpcion=1;
    for(String prop: listaPropiedades) {
        menuEP.put(numeroOpcion, prop); //opcion de menu, numero y nombre de propiedad
        numeroOpcion=numeroOpcion+1;
    }
    int salida=this.seleccionMenu(menuEP); // Método para controlar la elección correcta en el menú 
    return salida;
 
   }   

//Método para controlar la elección correcta de una opción en el menú que recibe como argumento   
private int seleccionMenu(Map<Integer,String> menu) {
   boolean valido; 
    int numero;
    String lectura;
    do { // Hasta que se hace una selección válida
      for(Map.Entry<Integer, String> fila : menu.entrySet()) {
            numero = fila.getKey();
            String texto = fila.getValue();
            this.mostrar(numero + " : " + texto);  // número de opción y texto
      }
      this.mostrar("\n Elige una opción: ");
      lectura = in.nextLine();  //lectura de teclado
      valido=this.comprobarOpcion(lectura, 0, menu.size()-1); //método para comprobar la elección correcta
    } while (!valido);
    return Integer.parseInt(lectura);
}

public ArrayList<String> obtenerNombreJugadores() { //método para pedir el nombre de los jugadores
    boolean valido; 
    String lectura;
    ArrayList<String> nombres = new ArrayList();
    do{ //repetir mientras que el usuario no escriba un número correcto 
        this.mostrar("Escribe el número de jugadores: (de 2 a 4):");
        lectura = in.nextLine();  //lectura de teclado
        valido=this.comprobarOpcion(lectura, 2, 4); //método para comprobar la elección correcta
    }while (!valido);
    
    for (int i = 1; i <= Integer.parseInt(lectura); i++) { //solicitud del nombre de cada jugador
      this.mostrar("Nombre del jugador " + i + ": ");
      nombres.add (in.nextLine());
    }
    return nombres;
}

//método para comprobar que se introduce un entero correcto, usado por seleccion_menu
private boolean comprobarOpcion(String lectura, int min, int max){ 
    boolean valido=true;   
    int opcion;
    try {  
         opcion =Integer.parseInt(lectura);
         if (opcion<min || opcion>max) { // No es un entero entre los válidos
              this.mostrar("el numero debe estar entre min y max");
               valido = false;}

    } catch (NumberFormatException e) { // No se ha introducido un entero
        this.mostrar("debes introducir un numero");
        valido = false;  
    }
    if (!valido) {
       this.mostrar("\n\n Seleccion erronea. Intentalo de nuevo.\n\n");
    }
    return valido;
}

 public void mostrar(String texto){ //método que muestra en pantalla el string que recibe como argumento
         
    System.out.println(texto);
}
 

}

File: QytetetRuby/lib/vista_textual_qytetet.rb
#encoding: utf-8


#require_relative "metodo_salir_carcel"

module InterfazTextualQytetet
class VistaTextualQytetet
  include ModeloQytetet
  def pausa
    puts "\nPulse Intro"
    gets.chomp
  end

  
  def seleccion_menu(menu)
     
    begin #Hasta que se hace una seleccionn valida
      valido= true
      for m in menu #se muestran las opciones del menuº
        mostrar( "#{m[0]}" + " : " + "#{m[1]}")
      end
      mostrar( "\n Elige un numero de opcion: ")
      captura = gets.chomp
      valido=comprobar_opcion(captura, 0, menu.length-1); #metodo para comprobar la eleccion correcta
      if (! valido) then
        mostrar( "\n\n ERROR !!! \n\n Seleccion erronea. Intentalo de nuevo.\n\n")
      end
    end while (! valido)
    Integer(captura)  

  end
  
  
  def comprobar_opcion(captura,min,max)
    # metodo para comprobar si la opcion introducida es correcta, usado por seleccion_menu
     valido=true
     begin
        opcion = Integer(captura)
        if (opcion<min || opcion>max) #No es un entero entre los validos
          valido = false
          mostrar('el numero debe estar entre min y max')
        end
      rescue Exception => e  #No se ha introducido un entero
        valido = false
        mostrar('debes introducir un numero')
     end 
     valido
  end
   
  
  def menu_gestion_inmobiliaria
    mostrar( 'Elige la gestion inmobiliaria que deseas hacer')
    menuGI=[[0, 'Siguiente Jugador'], [1, 'Edificar casa'], [2, 'Edificar Hotel'], [3, 'Vender propiedad'],[4, 'Hipotecar Propiedad'], [5, 'Cancelar Hipoteca']]
    salida=seleccion_menu(menuGI)
#    mostrar( 'has elegido')
#    mostrar(salida)
    return salida
  end
  
  def menu_salir_carcel
    mostrar('Elige el método para salir de la carcel')
    menuSC=[[0, 'Salir pagando'], [1, 'Tirar dado']]
    salida=seleccion_menu(menuSC)
#    mostrar( 'has elegido')
#    mostrar(salida)
    
    if(salida==0)
      metodo = MetodoSalirCarcel::PAGANDOLIBERTAD
    else
      metodo = MetodoSalirCarcel::TIRANDODADO
    end
    
    return metodo
  end
  
  def elegir_quiero_comprar
    mostrar('Elegir comprar título de propiedad')
    menuQC=[[0, 'Quiero comprar'], [1, 'No quiero comprar']]
    salida = seleccion_menu(menuQC)
    
    return salida==0
  end
  
 
        
  def menu_elegir_propiedad(lista_propiedades) # numero y nombre de propiedades            
      menuEP = Array.new
      menuEP<<[0, "    No elegir propiedad"]
      numero_opcion=1;
      for prop in lista_propiedades
          menuEP<<[numero_opcion, prop]; # opcion de menu, numero y nombre de propiedad
          numero_opcion=numero_opcion+1
      end
#      puts menuEP.inspect
      salida=seleccion_menu(menuEP); # Método para controlar la elección correcta en el menú 
      salida;
  end  
  
   def obtener_nombre_jugadores
      nombres=Array.new
      valido = true; 
      begin
        self.mostrar("Escribe el numero de jugadores: (de 2 a 4):");
        lectura = gets.chomp #lectura de teclado
        valido=comprobar_opcion(lectura, 2, 4); #método para comprobar la elección correcta
      end while(!valido)
    
      for i in 1..Integer(lectura)  #pide nombre de jugadores y los mete en un array
         mostrar('Jugador:  '+ i.to_s)
         nombre=gets.chomp
         nombres<<nombre
      end
    nombres
  end
  
  def mostrar(texto)  #metodo para mostrar el string que recibe como argumento
    puts texto
  end
  
  private :comprobar_opcion, :seleccion_menu
  
 
end
end

